src/Simplify_Add.cpp

```diff
@@ -112,6 +112,13 @@ Expr Simplify::visit(const Add *op, ExprInfo *bounds) {
              rewrite(y*x + y*z, y*(x + z)) ||
              rewrite(x*c0 + y*c1, (x + y*fold(c1/c0)) * c0, c1 % c0 == 0) ||
              rewrite(x*c0 + y*c1, (x*fold(c0/c1) + y) * c1, c0 % c1 == 0) ||
+
+             rewrite(slice(x, c0, c1, c2) + slice(y, c0, c1, c2), slice(x + y, c0, c1, c2)) ||
+             rewrite(slice(x, c0, c1, c2) + (z + slice(y, c0, c1, c2)), slice(x + y, c0, c1, c2) + z) ||
+             rewrite(slice(x, c0, c1, c2) + (slice(y, c0, c1, c2) + z), slice(x + y, c0, c1, c2) + z) ||
+             rewrite(slice(x, c0, c1, c2) + (z - slice(y, c0, c1, c2)), slice(x - y, c0, c1, c2) + z) ||
+             rewrite(slice(x, c0, c1, c2) + (slice(y, c0, c1, c2) - z), slice(x + y, c0, c1, c2) - z) ||
```

Comment:
Are these always valid via a cost model for shuffling (across all backends)? For example, suppose the stride was 8 (and lanes also 8) - wouldn't the sum of length-8 vectors + 2 shuffles be cheaper than a sum of length-64 vectors + 1 shuffle?

Reply:
(same question for the Min/Max/Mul/Sub rules below)

Reply:
If stride is 8 and lanes is 8, then you're extracting one element from each vector, so the 2 shuffles are actually 16 shuffles. It's 16 shuffles and one add vs 8 adds and 8 shuffles.

Reply:
More generally, for any binary single-cycle op that consumes two inputs, it should be better to hoist the slice.

Reply:
Okay cool. Sgtm